### PR TITLE
Optimize the number of salt calls on minion startup

### PR DESCRIPTION
**Pull request by nadvornik**
## What does this PR change?

This PR  reduces the number of salt calls on minion registration and startup.
It merges multiple salt call during registration into one call `state.apply util.systeminfo_full`.

The new workflows are following:

Standard minion registration:
1. minion start event with startup grains
2. sync call to `state.apply util.systeminfo_full`
3. sync call for distro-specific product identification
4. async call to set ceritficates, channels, packages
5. async call to update package list
6. async call to update hardware info

Registration of saltboot minion:
1. minion start event with startup grains (with `saltbootInitrd=true`)
2. sync call to `state.apply util.systeminfo_full`
3. async call to `state.apply saltboot`
4. minion start event with startup grains  (with `saltbootInitrd=false`)
5. async call to `state.apply util.systeminfo`
6. image deployed event triggers the rest of registration:
7. sync call for distro-specific product identification
8. async call to set ceritficates, channels, packages
9. async call to update package list
10. async call to update hardware info

Reboot of standard minion:
1. minion start event with startup grains
2. async call to `state.apply util.systeminfo`

Reboot of satboot minion:
1. minion start event with startup grains (with `saltbootInitrd=true`)
2. async call to `state.apply saltboot`
3. minion start event with startup grains  (with `saltbootInitrd=false`)
4. async call to `state.apply util.systeminfo`

Typically, a failure of `state.apply util.systeminfo_full` is the only fatal failure, which can abort the registration. On other failures, the registration is finished and the error is visible in event log.

This PR drops `MinionStartEventMessage`, all the functionality is moved to `RegisterMinionEventMessageAction.applyMinionStartStates`.

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: not an user visible change

- [x] **DONE**

## Test coverage
- Unit tests were added

- [x] **DONE**

## Links

Part of fix for https://github.com/SUSE/spacewalk/issues/19063
Obsoletes https://github.com/uyuni-project/uyuni/pull/6087

- [ ] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests" 
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java [hunk 1]
```diff
@@ -408,45 +417,38 @@ private void updateKickStartSession(Optional<ActivationKey> activationKey) {
...
                                            Optional<String> activationKeyOverride,
-                                           boolean isSaltSSH) {
+                                           boolean isSaltSSH,
+                                           boolean saltbootInitrd) {
         Optional<User> creator = MinionPendingRegistrationService.getCreator(minionId);
-        // take care that custom grains are deployed on the minion before we request them
-        saltApi.syncGrains(new MinionList(minionId));
-        ValueMap grains = new ValueMap(saltApi.getGrains(minionId).orElseGet(HashMap::new));
-        MinionServer minion = migrateOrCreateSystem(minionId, isSaltSSH, activationKeyOverride, machineId, grains);
+        Org org = OrgFactory.getSatelliteOrg();
```
Comment (aaannz) [inline java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java:430]: Why is this here and not later with proxy detection. I can't see it used before the second declaration.

File: java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java [hunk 2]
```diff
@@ -533,19 +539,26 @@ else if (!minion.getOrg().equals(org)) {
...
                         minionId);
-                prepareRetailMinionForSaltboot(minion, org, grains);
-                applySaltboot(minion);
+                try {
+                    prepareRetailMinionForSaltboot(minion, org, grains);
+                    applySaltboot(minion, false);
+                }
+                catch (Throwable t) {
+                    LOG.error("Error preparing saltboot configuration for minion id: {}", minionId, t);
+                    // do not throw, leave the minion as registered in initrd
```
Comment (aaannz) [inline java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java:551]: This is just log error, but at this moment we already have server entry in database and by not throwing this entry will not be removed. Can we add failed event to the event log as well?

File: java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java (hunk 1 above)
Comment (aaannz) [reply java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java:430]: Ah, its needed for catch later. But org in exception will be default org not related to the org of the minion. This doesn't seem right. I think org parameter can be removed as minionID has to be unique across all orgs anyway.

File: java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java (hunk 2 above)
Comment (aaannz) [reply java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java:551]: Ah, nevermind. They are added in the methods themselves.

**Comment by mcalmer** [comment]
@nadvornik please check what happens when no startup grains are configured.

**Comment by nadvornik** [comment]
> @nadvornik please check what happens when no startup grains are configured.

This triggers an extra call to saltApi.getGrains in RegisterMinionEventMessageAction.execute. I did not touch this part.

**Review by cbosdo — APPROVED** [review]
LGTM... though I clearly don't have the full bootstrap picture in mind

**Comment by cbosdo** [comment]
@nadvornik you will need to add an entry in both `spacewalk-java.changes` and `susemanager-sls.changes`

**Review by aaannz — APPROVED** [review]

**Pull request merged**